src/Lean/PrettyPrinter/Delaborator/Builtins.lean

```diff
@@ -137,15 +157,24 @@ def delabAppImplicit : Delab := whenNotPPOption getPPExplicit do
       let paramKinds ← liftM (getParamKinds fn <|> pure #[])
       pure (stx, paramKinds.toList, #[]))
     (fun (fnStx, paramKinds, argStxs) => do
-      let arg ← getExpr;
-      let implicit : Bool := match paramKinds with -- TODO: check why we need `: Bool` here
-        | [ParamKind.implicit (some v)] => !v.hasLooseBVars && v == arg
-        | ParamKind.implicit none :: _  => true
-        | _                             => false
-      if implicit then
-        pure (fnStx, paramKinds.tailD [], argStxs)
-      else do
+      let arg ← getExpr
+      let opts ← getOptions
+      let argKind : ImplicitArgKind := match paramKinds with
+        | [ParamKind.implicit _ (some v)] => 
+          if !v.hasLooseBVars && v == arg then ImplicitArgKind.skip else ImplicitArgKind.provide
+        | ParamKind.implicit name none :: _  => do
+          if name == `motive && showMotive arg opts
+          then ImplicitArgKind.name name                        
+          else ImplicitArgKind.skip
+        | _ => ImplicitArgKind.provide
+      match argKind with
```

Comment:
I'd suggest to fuse this `match` with the prior `let`, getting rid of `ImplicitArgKind`, which doesn't seem to be used anywhere else. You can also factor out the common `(fnStx, paramKinds.tailD [],` part, then the `skip` and `provide` cases are trivial enough that duplicating them should not be an issue.

Reply:
I removed `ImplicitArgKind` at your suggestion. Now the `let` binds an `Option Syntax`, we match on this to compute the new `argStxs` and then all paths return together.